app/models/sendgrid_event.rb

```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal: true
+class SendgridEvent < ApplicationRecord
+
+  store_accessor :payload, :email
+
+  # To make allowances for occasional inbox down time, this counts a maximum of one fail per day,
+  # e.g.:
+  #
+  # - If 2 emails are sent on the same day, and both fail, then this counts as only 1 fail, not
+  #   2 fails.
+  #
+  # - If 2 emails are sent on 2 separate days, and both fail, then this counts as 2
+  #   fails, as the failures were on different days.
+  def self.fails_since_last_delivery(email)
+    last_delivered_at = where_email(email).where_event_type("delivered").maximum(timestamp_sql)
+
+    fails_query =
+      select("DISTINCT(date_trunc('day', #{timestamp_sql}))").where_email(email).where_event_type("dropped", "bounce")
+
+    fails_query = fails_query.where("#{timestamp_sql} > ?", last_delivered_at) if last_delivered_at
+
+    fails_query.count
+  end
+
+  def self.where_email(email)
+    where("(payload->>'email') = ?", email)
+  end
+
+  def self.where_event_type(*event_types)
+    where("(payload->>'event') IN (?)", event_types)
+  end
+
+  def self.where_sg_event_id(sg_event_id)
+    where("(payload->>'sg_event_id') = ?", sg_event_id)
+  end
+
+  def self.timestamp_sql
+    "to_timestamp((payload->>'timestamp')::double precision)"
+  end
+
+  def self.process_later(payload)
+    return if where_sg_event_id(payload[:sg_event_id]).exists?
+
+    transaction do
+      event = create!(payload: payload, status: "pending")
+      self.delay.process(event.id)
```

Comment:
What is the upside of using a seperate job to update `mail_fails`?

Reply:
In short, fault tolerance. Its to confirm webhook event delivery with a 200 OK response back to the client/SendGrid asap whilst minimising the likelihood of responding with an error code. When there are any processing errors, these happen in the job, the client/Sendgrid doesn't have to retry delivery, and we're not coupled to SendGrid's retry behaviour. The app (via delayed job) is in control of retries.

Another reason is if event processing became an intensive workload which takes longer than the request timeout threshold, though this is premature optimization, as the workload is light currently, experience nudges me to write webhook endpoints like this (offloading to a job)  from the start, as the workload often grows intensive.